filecoin/activeretrievals.go

```diff
@@ -0,0 +1,252 @@
+		return
+	}
+	ar.queriesFinished++
+	log.Debugf("QueryCandidatedFinished for %s (%d active, %d/%d query candidates, %d/%d retrieval candidates)", retrievalCid, len(arm.arMap), ar.queriesFinished, ar.queryCandidateCount, ar.retrievalsFinished, ar.retrievalCandidateCount)
+	arm.maybeFinish(retrievalCid, ar)
+}
+
+// RetrievalCandidatedFinished registers that a retrieval has finished with one
+// of the query candidates. In the case of a failure, it's used to increment the
+// number of finished so we can check the finished count against the number of
+// expected candidates for possible clean-up. In the case of success, we can
+// assume no more retrievals will occur so we can jump straight to clean-up.
+func (arm *ActiveRetrievalsManager) RetrievalCandidatedFinished(retrievalCid cid.Cid, success bool) {
+	arm.lk.Lock()
+	defer arm.lk.Unlock()
+
+	ar := arm.findActiveRetrievalFor(retrievalCid)
+	if ar == nil {
+		log.Errorf("Unexpected active retrieval RetrievalCandidatedFinished for %s (%d active, %d/%d query candidates, %d/%d retrieval candidates)", retrievalCid, len(arm.arMap), ar.queriesFinished, ar.queryCandidateCount, ar.retrievalsFinished, ar.retrievalCandidateCount)
+		return
+	}
+	ar.retrievalsFinished++
+	if success {
+		// if success, there won't be any more retrieval attempts so we can short-circuit clean-up
+		ar.retrievalCandidateCount = ar.retrievalsFinished
+	}
+	log.Debugf("RetrievalCandidatedFinished for %s (%d active, %d/%d query candidates, %d/%d retrieval candidates)", retrievalCid, len(arm.arMap), ar.queriesFinished, ar.queryCandidateCount, ar.retrievalsFinished, ar.retrievalCandidateCount)
+	arm.maybeFinish(retrievalCid, ar)
+}
+
+// findActiveRetrievalFor looks up a retrieval in the map by CID but will also
+// search for rootCid matches as a secondary option since they can also be
+// identified by that way.
+// Note that one implication of looking up a CID in both places is that calls to
+// New() will return ErrRetrievalAlreadyRunning if that CID is involved as
+// either a primary CID or a root CID. But this is reasonable since a successful
+// retrieval will yield the newly requested CID regardless.
+func (arm *ActiveRetrievalsManager) findActiveRetrievalFor(cid cid.Cid) *activeRetrieval {
+	ar, ok := arm.arMap[cid]
+	if !ok {
+		for _, ar := range arm.arMap {
+			if cid.Equals(ar.rootCid) {
+				return ar
+			}
+		}
+	}
+	return ar
+}
+
+// maybeFinish checks whether we've completed all expected queries and
+// retrievals, and if so, cleans up the record. "complete" means that we've
+// recorded a finish for the expected number of queries and retrievals.
+// This function expects the lock to be held by the caller!
+func (arm *ActiveRetrievalsManager) maybeFinish(retrievalCid cid.Cid, ar *activeRetrieval) {
+	if ar.queriesFinished >= ar.queryCandidateCount && ar.retrievalsFinished >= ar.retrievalCandidateCount {
```

Comment:
OK I think I'm seeing a potential memory leak here...
the retrieval attempts run until there is a success, but we may simply never try candidates if we have a successful retrieval. The way your code is written in retriever.go, you set the candidate count after the query phase is over, and called RetrievalCandidateFinished in the events. But I don't think you will get a RetrievalCandidateFinished for any candidates that aren't tried, so you will never delete from the map.

Reply:
as discussed on call - the bool in `RetrievalCandidatedFinished()` accounts for this by short-circuiting the clean-up on a successful retrieval - essentially "we got one, wrap it up"